pkg/debug/debug.go

```diff
@@ -0,0 +1,144 @@
+package debug
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/spf13/afero"
+)
+
+const (
+	testArtifactsDirectoryVarName = "TEST_ARTIFACTS_DIRECTORY"
+)
+
+var (
+	// globals for overriding in unit tests
+	artifactsDirectoryBase = os.Getenv(testArtifactsDirectoryVarName) // nolint:gochecknoglobals
+	execCommand            = exec.Command                             // nolint:gochecknoglobals
+)
+
+// CollectArtifacts collects useful debugging artifacts from a given namespace.
+// Should typically be called if CurrentGinkgoTestDescription().Failed, like this:
+//   debug.CollectArtifacts(afero.NewOsFs(), GinkgoWriter, TestNamespace, KubeConfigPath, KubectlPath)
+func CollectArtifacts(fs afero.Fs, writer io.Writer, namespace, kubeConfigPath, kubectlPath string) {
+	err := collectNamespacedResources(fs, writer, namespace, kubeConfigPath, kubectlPath)
+	if err != nil {
+		_, _ = fmt.Fprintf(writer, "collection of resources for debugging failed: %v\n", err)
+	}
+}
+
+func collectNamespacedResources(fs afero.Fs, writer io.Writer, namespace, kubeConfigPath, kubectlPath string) error {
+	if artifactsDirectoryBase == "" {
+		return fmt.Errorf("$%s not set", testArtifactsDirectoryVarName)
+	}
+
+	_, _ = fmt.Fprintf(writer, "collecting namespaced resources for debugging...\n")
+
+	cmd := execCommand(
+		kubectlPath,
+		"--kubeconfig",
+		kubeConfigPath,
+		"api-resources",
+		"--verbs=list",
+		"--namespaced=true",
+		"-o",
+		"name")
+	cmd.Stderr = writer
+
+	output, err := cmd.Output()
+	if err != nil {
+		return fmt.Errorf("fetching API resource types failed: %v", err)
+	}
+
+	artifactsDirectory := path.Join(artifactsDirectoryBase, namespace)
+
+	err = fs.MkdirAll(artifactsDirectory, 0777)
+	if err != nil {
+		return fmt.Errorf("creating %q failed: %v", artifactsDirectory, err)
+	}
+
+	resources := strings.Split(string(output), "\n")
+	groupedResources := make(map[string][]string)
+
+	for _, resources := range resources {
+		if resources == "" {
+			continue
+		}
+
+		resourcesAndGroup := strings.SplitN(resources, ".", 2)
+
+		var group string
+
+		if len(resourcesAndGroup) == 2 {
+			group = resourcesAndGroup[1]
+		}
+
+		groupedResources[group] = append(groupedResources[group], resources)
+	}
+
+	var wg sync.WaitGroup
+
+	for group, resourcesSlice := range groupedResources {
+		sort.Strings(resourcesSlice)
+
+		resources := strings.Join(resourcesSlice, ",")
+
+		var fileName string
+
+		if group == "" {
+			fileName = "resources.yaml"
+		} else {
+			fileName = fmt.Sprintf("resources-%s.yaml", group)
+		}
+
+		wg.Add(1)
+
+		go collectResources(fs, writer, &wg, namespace, fileName, resources,
```

Comment:
Is there a reason other than possible performance gains to do this in parallel? Are there performance issues when running this in serial? This looks like premature optimization and there could be issues using `afero.Fs` concurrently... are `fs` and `writer` thread-safe? Let's not use go-routines here unless there's a good reason to do so. And if there is a good reason to do so, use channels to synchronize writing to `fs` and `writer`.

Reply:
See point 4 in https://github.com/kudobuilder/test-tools/pull/24#issuecomment-607104384